sympy/polys/rings.py

```diff
@@ -2189,8 +2192,30 @@ def _gcd(f, g):
             return f._gcd_QQ(g)
         elif ring.domain.is_ZZ:
             return f._gcd_ZZ(g)
-        else: # TODO: don't use dense representation (port PRS algorithms)
-            return ring.dmp_inner_gcd(f, g)
+        else:
+            old_gcd = ring.dmp_inner_gcd(f, g)
+            if not ring.domain.is_Field:
+                h = _gcd_prs(f, g)
+                cff = f.quo(h)
+                cfg = g.quo(h)
+                new_gcd = h, cff, cfg
+                if new_gcd != old_gcd:
+                    raise RuntimeError(f"GCD is different for\n\n"
+                    f"f:\n{f} :\n\ng:\n{g}\n\nOld GCD = {old_gcd}\n"
+                    f"New GCD = {new_gcd}")
+                else:
+                    return new_gcd
+            else:
+                h = _gcd_prs(f, g).monic()
+                cff = f.quo(h)
+                cfg = g.quo(h)
+                new_gcd = h, cff, cfg
+                if new_gcd != old_gcd:
+                    raise RuntimeError(f"GCD is different for\n\n"
+                    f"f:\n{f} :\n\ng:\n{g}\n\nOld GCD = {old_gcd}\n"
+                    f"New GCD = {new_gcd}")
+                else:
+                    return new_gcd
```

Comment:
It would be better to make a separate method `_gcd_prs` so this just looks like:
```python
        # TODO: implement better algorithm selection
        if ring.domain.is_QQ:
            return f._gcd_QQ(g)
        elif ring.domain.is_ZZ:
            return f._gcd_ZZ(g)
        else:
            return f._gcd_prs(g)
```